backend/src/stwo/circuit_builder.rs

```diff
@@ -0,0 +1,236 @@
+        self.log_n_rows + 1
+    }
+    fn evaluate<E: EvalAtRow>(&self, mut eval: E) -> E {
+
+            let mut witness_eval = Vec::with_capacity(self.witness.unwrap().len());
+            for _ in 0..self.witness.unwrap().len() {
+                witness_eval.push(eval.next_interaction_mask(0,[0,1] ));
+    
+            }
+            println!("This is the witness eval {:?}", self.witness);
+            
+        
+        // Add polynomial identities
+        let identities = self
+            .analyzed
+            .identities_with_inlined_intermediate_polynomials()
+            .into_iter()
+            .filter(|id| id.kind == IdentityKind::Polynomial)
+            .collect::<Vec<_>>();
+        
+        if !identities.is_empty() {
+            identities.iter().for_each(|id| {
+                let expr = id.expression_for_poly_id();
+                //let name = id.to_string();
+                println!("This is the expression {:?}", expr);
+                let expr = to_stwo_expression(&self.witness_columns,expr, &witness_eval,&eval);
+                eval.add_constraint(expr);
+            });
+        }
+        eval
+    }
+}
+
+fn to_stwo_expression<T: FieldElement, E: EvalAtRow>(
+    witness_columns:&BTreeMap<PolyID, usize>,
+    expr: &AlgebraicExpression<T>,
+    witness_eval: &Vec<[<E as EvalAtRow>::F;2]>,
+    eval:  &E,
+) -> E::F {
+    match expr {
+        AlgebraicExpression::Number(n) => {
+            E::F::one()
+        }
+        AlgebraicExpression::Reference(polyref) => {
+            let poly_id = polyref.poly_id;
+            let interaction = match polyref.next {
+                false => {
+                    let index = witness_columns[&poly_id];
+            witness_eval[index][0].into()
+
+                },
+                true => {
+                   
+                    let index = witness_columns[&poly_id];
+                    println!("witness_eval[index][1].into() is {:?}",witness_eval[index][1]);
+                    witness_eval[index][1].into()
+                }
+            };
+            interaction
+        }
+        AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation {
+            left: lhe,
+            op,
+            right: powdr_rhe,
+        }) => {
+            let lhe = to_stwo_expression(witness_columns,lhe, witness_eval,eval);
+            let rhe = to_stwo_expression(witness_columns,powdr_rhe, witness_eval,eval);
+            match op {
+                AlgebraicBinaryOperator::Add => {
+                    lhe + rhe
+                }
+                AlgebraicBinaryOperator::Sub => {
+                    lhe - rhe
+                }
+                AlgebraicBinaryOperator::Mul => {
+                    lhe * rhe
+                }
+                AlgebraicBinaryOperator::Pow => {
+                    let AlgebraicExpression::Number(e) = powdr_rhe.as_ref() else {
+                        panic!("Expected number in exponent.")
+                    };
```

Comment:
I think plonky3 supports more complex expressions here, we should check there.

Reply:
now the to_stwo_expression() function follows the  implementation in plonky3